### PR TITLE
Text to speech word tracking events

### PR DESCRIPTION
**Pull request by gabrieljoelc**
# Why

We need the ability to highlight each word of a string of text in a native UI being uttered via text-to-speech SDK's. The current expo-speech does not implement the related events for Android and iOS.

# How

I added the relevant events for Android and iOS as shown in their docs, respectively, to the existing expo-speech module:
- [Android `onRangeStart`](https://developer.android.com/reference/android/speech/tts/UtteranceProgressListener#onRangeStart(java.lang.String,%20int,%20int,%20int))
- [iOS delegate with `willSpeakRangeOfSpeech`](https://developer.apple.com/documentation/avfaudio/avspeechsynthesizerdelegate/1619681-speechsynthesizer/)

# Test Plan

I smoke tested this in the application I need it for on a physical iPhone and Android emulator. The e2e tests for Speech are a bit rough at this time, so I was afraid to touch them.

Here's a rough sketch of how it works in an application (see an [early prototype](https://github.com/gabrieljoelc/TextToSpeech/pull/2)):

```js
function onWillSayNextString({location, length}) {
  console.log('onWillSayNextString character options', { location, length });
}

function customSpeakPressHandler() {
  speak('Some text to utter', { onWillSayNextString });
}

return (
  <Button title='Speak' onPress={customSpeakPressHandler} />
);
```

# Checklist

- [x] Working through https://github.com/expo/expo/blob/main/CONTRIBUTING.md
- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

I couldn't figure this one out and the contribution documentation didn't mention this step. I tried running it in `packages/expo-speech` but it said "The android and ios projects are malformed, would you like to clear the project files and reinitialize them?" I didn't change the project structures for this PR and wasn't sure what this meant.
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-speech/CHANGELOG.md
Comment (barthap) [inline packages/expo-speech/CHANGELOG.md:18]: This should be in the `## Unpublished` section

File: apps/native-component-list/package.json
Comment (barthap) [inline apps/native-component-list/package.json:118]: Please do not bump versions manually, we have a separate package publishing process and this is not a scope of this PR
```suggestion
    "expo-speech": "~11.0.0",
```

File: packages/expo/CHANGELOG.md
Comment (barthap) [inline packages/expo/CHANGELOG.md:10]:
```suggestion
```
Remove this. Bot suggested this changelog, because you edited files inside `packages/expo/`. Revert these version bumps and the bot will be ok with the changelog entry only in `packages/expo-speech/CHANGELOG.md`

File: packages/expo-speech/android/src/main/java/expo/modules/speech/SpeechModule.kt
Comment (barthap) [inline packages/expo-speech/android/src/main/java/expo/modules/speech/SpeechModule.kt:148]: Are these positions correct?
Shouldn't location be determined by `start` and length by `end - start`?

According to Android docs, the `frame` means audio frame, not the character position in string:
> `frame` - int: The position in frames in the audio of the request where this range is spoken.

**Review by barthap — CHANGES_REQUESTED** [review]
Wow, this feature is awesome! Thanks for your contribution!

The logic of passing events Native->JS looks good. Please revert all these version bumps and we're close to merging.

File: apps/native-component-list/package.json
Comment (gabrieljoelc) [reply apps/native-component-list/package.json:118]: ok my bad. the builds were failing without this so I thought I had to do it. I'll revert.

File: packages/expo/CHANGELOG.md
Comment (gabrieljoelc) [reply packages/expo/CHANGELOG.md:10]: oh i see. my bad. will do.

File: packages/expo-speech/android/src/main/java/expo/modules/speech/SpeechModule.kt
Comment (gabrieljoelc) [reply packages/expo-speech/android/src/main/java/expo/modules/speech/SpeechModule.kt:148]: hm, you are right. I did smoke test but I think I know what happened. I smoke tested before I got the highlighting test implemented and didn't test again. I will fix.

File: packages/expo-speech/CHANGELOG.md
Comment (barthap) [reply packages/expo-speech/CHANGELOG.md]: If you have any ideas on how to make the bot suggestion less confusing, let us know 😉 

cc @tsapeta

File: packages/expo-speech/android/src/main/java/expo/modules/speech/SpeechModule.kt
Comment (barthap) [reply packages/expo-speech/android/src/main/java/expo/modules/speech/SpeechModule.kt]: I see that `expo-speech` has `onBoundary` undocumented. Looking at [web MDN docs](https://developer.mozilla.org/en-US/docs/Web/API/SpeechSynthesisUtterance/boundary_event), it technically represents the same, so this might be a good idea, but it requires some more work with unifying the API and documenting it:
- Text position: iOS: ✅ , android: ✅ web: ❌ 
- Text length: iOS: ✅ , android: ✅ web: ❌ 
- Sound timestamp: iOS: ❌  , android: ✅ web: ✅

I personally have no strong opinion on this

Comment (gabrieljoelc) [reply packages/expo-speech/android/src/main/java/expo/modules/speech/SpeechModule.kt]: yes, I have been looking into those.


> Text position: iOS: ✅ , android: ✅ web: ❌

This does already exist as [`charIndex`](https://developer.mozilla.org/en-US/docs/Web/API/SpeechSynthesisEvent/charIndex).

> Text length: iOS: ✅ , android: ✅ web: ❌

Length does not exist in the [MDN docs](https://developer.mozilla.org/en-US/docs/Web/API/SpeechSynthesisUtterance/boundary_event) but [`charLength`](https://github.com/microsoft/TypeScript/blob/main/lib/lib.dom.d.ts#L1719) does exist in this [type](https://github.com/microsoft/TypeScript/blob/main/lib/lib.dom.d.ts#L1717-L1723) that my IDE is able to find when I "go to definition" (a RN shim?).

> Sound timestamp: iOS: ❌ , android: ✅ web: ✅

For Android, are you referring to the [`frame`](https://developer.android.com/reference/android/speech/tts/UtteranceProgressListener#onRangeStart(java.lang.String,%20int,%20int,%20int)) parameter?

Here's what I'm considering for now:

```js
// Speech.js
setSpeakingListener('Exponent.speakingWillSayNextString', ({ id, charIndex, charLength, name }) => {
    const options = _CALLBACKS[id];
    if (options && options.onBoundary) {
        options.onBoundary({
            charIndex,
            charLength,
            name,
        });
    }
});
```

```kt
override fun onRangeStart(utteranceId: String, start: Int, end: Int, frame: Int) {
  var map = Bundle().apply {
    putString("id", utteranceId)
    putInt("charIndex", start)
    putInt("charLength", end - start)
    putInt("name", "onRangeStart")
  }
  emitter.emit("Exponent.speakingWillSayNextString", map)
}
```

```objective-c
- (void)speechSynthesizer:(AVSpeechSynthesizer *)synthesizer
  willSpeakRangeOfSpeechString:(NSRange)characterRange utterance:(nonnull AVSpeechUtterance *)utterance
{
  id<EXEventEmitterService> emitter = [_moduleRegistry getModuleImplementingProtocol:@protocol(EXEventEmitterService)];
  if (emitter != nil) {
    [emitter sendEventWithName:@"Exponent.speakingWillSayNextString" body:@{ @"id": ((EXSpeechUtteranceWithId *) utterance).utteranceId, @"charIndex": @(characterRange.location), @"charLength": @(characterRange.length), @"name": "willSpeakRangeOfSpeechString" }];
}
```

I'm on the fence about sending [`name`](https://developer.mozilla.org/en-US/docs/Web/API/SpeechSynthesisEvent/name) and can remove if the semantics don't match for the native events, respectively.

Thoughts?

Comment (gabrieljoelc) [reply packages/expo-speech/android/src/main/java/expo/modules/speech/SpeechModule.kt]: Here's a [proposal](https://github.com/paper-co/expo/pull/3). I decided against using `name` or worrying about the sound timestamp parameters for now. I added caveats to the documentation similar to something I found [here](https://github.com/expo/expo/blob/main/packages/expo-device/src/Device.ts#L264-L280).

File: packages/expo-speech/src/Speech/Speech.types.ts
Comment (barthap) [reply packages/expo-speech/src/Speech/Speech.types.ts:51]: Also, correct me if I'm wrong but web [`SpeechSynthesisEvent`](https://developer.mozilla.org/en-US/docs/Web/API/SpeechSynthesisEvent) doesn't have the `charLength` property so this would cause typescript errors.

I'm wondering if having both web `SpeechEventCallback` and your previous callback `({ charIndex:  number, charLength: number }) => void` a union type would make sense:
```ts
/**
 * Your description here from previous commits
 * @platform ios
 * @platform android
 */
export type NativeBoundaryEventCallback = (({ charIndex:  number, charLength: number }) => void;

// then in options
onBoundary?: NativeBoundaryEventCallback | SpeechEventCallback | null;
```
There are already union types for other options like `onStart`. What do you think? 🤔

**Review by barthap — CHANGES_REQUESTED** [review]
Linter is failing (check CI)
You can run `yarn lint --fix` from inside the `expo-speech` directory - this should do the job

Besides that, the code looks good now, at first glance this shouldn't conflict with the web version, nice job!
One downside is that we don't have good typing for this (see comment) because of web version

File: packages/expo-speech/src/Speech/Speech.types.ts
Comment (gabrieljoelc) [reply packages/expo-speech/src/Speech/Speech.types.ts:51]: > web [SpeechSynthesisEvent](https://developer.mozilla.org/en-US/docs/Web/API/SpeechSynthesisEvent) doesn't have the charLength property

yes, I mentioned this in https://github.com/expo/expo/pull/20726#discussion_r1068467834, and that it "does exist in this [type](https://github.com/microsoft/TypeScript/blob/main/lib/lib.dom.d.ts#L1717-L1723) that my IDE is able to find when I "go to definition" (a RN shim?)." know what that type is about?

yes that type change makes sense. i'll do that

Comment (barthap) [reply packages/expo-speech/src/Speech/Speech.types.ts]: I have no idea where this type comes from either 🤷 but I'd stick to the MDN docs where `charLength` isn't mentioned

Comment (gabrieljoelc) [reply packages/expo-speech/src/Speech/Speech.types.ts]: Added commits with the union type and ran `yarn lint --fix`

**Review by barthap — APPROVED** [review]
CI passed this time 🎉 
Looks good to me, but I'll let other reviewers take a look before merging.

**Review by lukmccall — APPROVED** [review]
Looking great 🔥
Thanks for your contribution!

**Pull request merged**
